### PR TITLE
feat(provider): add SAML passport authentication strategy

### LINKED ISSUE #128: Add SAML passport authentication strategy
## Description
New strategy will be added for SAML authentication similar to previous strategies that have been added before.

### PR DESCRIPTION
**Pull request by RaghavaroraSF**
## Description

feat(provider): add SAML passport authentication strategy

Added  passport authentication srategy For SAML in src.
Added types, component, enums and keys related to it.
Added relevant test cases.

Fixes #128

## Type of change

Please delete options that are not relevant.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Intermediate change (work in progress)

## How Has This Been Tested?

Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration

- [ ] Test A
- [ ] Test B

## Checklist:

- [ ] Performed a self-review of my own code
- [ ] npm test passes on your machine
- [ ] New tests added or existing tests modified to cover all changes
- [ ] Code conforms with the style guide
- [ ] API Documentation in code was updated
- [ ] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

**Comment by yeshamavani** [comment]
Link this PR with a issue.

File: src/strategy-adapter.ts [hunk 1]
```diff
@@ -61,7 +61,7 @@ export class StrategyAdapter<T> {
 
       strategy.redirect = (url: string) => {
         if (response) {
-          response.redirect(url, 302);
+          response.redirect(302, url);
```
Comment (yeshamavani) [inline src/strategy-adapter.ts:64]: why this change. does this affect other strategies?

File: src/strategies/types/types.ts [hunk 2]
```diff
@@ -115,6 +116,15 @@ export namespace VerifyFunction {
       req?: Request,
     ): Promise<T | null>;
   }
+  export interface SamlFn<T = IAuthUser> extends GenericAuthFn<T> {
+    (
+      // accessToken: string,
+      // refreshToken: string,
```
Comment (yeshamavani) [inline src/strategies/types/types.ts:122]: Why is this commented

File: src/strategies/SAML/saml-strategy-factory-provider.ts [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+        const user = await verifyFn(profile, cb, req);
+        if (!user) {
+          throw new HttpErrors.Unauthorized(AuthErrorKeys.InvalidCredentials);
+        }
+        cb(null, user as unknown as Record<string, unknown>);
+      } catch (err) {
+        cb(err);
+      }
+    };
+    if (options && options.passReqToCallback === true) {
```
Comment (yeshamavani) [inline src/strategies/SAML/saml-strategy-factory-provider.ts:53]: what is passReqToCallback is false

**Review by yeshamavani — CHANGES_REQUESTED** [review]

File: src/strategies/types/types.ts (hunk 2 above)
Comment (RaghavaroraSF) [reply src/strategies/types/types.ts:122]: For SAML only 2 arguments are needed that's why it was commented.

File: src/strategies/SAML/saml-strategy-factory-provider.ts (hunk 3 above)
Comment (RaghavaroraSF) [reply src/strategies/SAML/saml-strategy-factory-provider.ts:53]: Resolved
